### PR TITLE
Use "for_each" syntax clipping grad norm

### LINKED ISSUE #92736: [FSDP] Add `foreach` support to `FSDP.clip_grad_norm_()`
See https://github.com/pytorch/pytorch/pull/91846 for `torch.nn.utils.clip_grad_norm_()` changes. Since `FSDP.clip_grad_norm_()` has to re-implement the logic to include gradient unsharding, we have to manually upstream improvements like this.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501

### PR DESCRIPTION
**Pull request by thomasw21**
- Related to: #102429  
- Fixes: #92736

### CODE REVIEW

**Comment by awgu** [comment]
Feel free to close https://github.com/pytorch/pytorch/issues/92736 if this PR lands!

File: aten/src/ATen/native/cuda/ForeachReduceOp.cu [hunk 1]
```diff
@@ -117,10 +117,16 @@ std::vector<Tensor> foreach_tensor_norm_cuda(TensorList tensors, const Scalar& o
       const auto scalar_type = t.scalar_type();
       return at::isIntegralType(scalar_type, /*includeBool*/true) || at::isComplexType(scalar_type);
   });
+  const bool is_same_dtype = dtype.has_value() ? std::any_of(tensors.begin(), tensors.end(), [&](const auto & t) {
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/ForeachReduceOp.cu:120]: Is this not already checked in check_foreach_api_restrictions?

Should this be checked there instead?

Comment (thomasw21) [reply aten/src/ATen/native/cuda/ForeachReduceOp.cu:120]: Sorry perhaps a naming issue, but `is_same_dtype` checks that all tensors are of the same dtype as the specified `dtype` passed in argument (if specified). Typically you can run `torch.linalg.vector_norm(..., dtype=torch.bfloat16)`. It doesn't just check that all tensors are of the same dtype.

**Comment by thomasw21** [comment]
Gentle ping @janeyx99 ! Let me know if there are things I still need to change?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -2032,18 +2038,39 @@ def _get_grad_norm(
...
+            )
+        else:
+            grad_norms.extend(
+                [
+                    torch.linalg.vector_norm(g, norm_type, dtype=torch.float32)
+                    for g in grads
+                ]
+            )
+
     grad_norm = torch.linalg.vector_norm(
```
Comment (janeyx99) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2068]: Wait...isn't this vector_norm call redundant after the for loop?

Comment (thomasw21) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2068]: Not sure what you mean, you run vector_norm once per param and then aggregate the results for all params by running vector_norm on the list of norms.

Comment (janeyx99) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2068]: Ah, sorry for the misunderstanding

File: aten/src/ATen/native/cuda/ForeachReduceOp.cu [hunk 3]
```diff
@@ -134,9 +135,18 @@ std::vector<Tensor> foreach_tensor_norm_cuda(
         return at::isIntegralType(scalar_type, /*includeBool*/ true) ||
             at::isComplexType(scalar_type);
       });
-  if (!can_use_fast_route(tensors) || has_int_or_complex ||
+  const bool is_same_dtype = dtype.has_value()
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/ForeachReduceOp.cu:138]: check_fast_path_restrictions already makes this check (which is called from can_use_fast_route, I believe) so this check shouldn't be necessary

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -10604,7 +10604,7 @@
     CUDA: foreach_tensor_neg_cuda_
   autogen: _foreach_neg.out
 
-- func: _foreach_norm.Scalar(Tensor[] self, Scalar ord=2) -> Tensor[]
+- func: _foreach_norm.Scalar(Tensor[] self, Scalar ord=2, ScalarType? dtype=None) -> Tensor[]
```
Comment (janeyx99) [inline aten/src/ATen/native/native_functions.yaml:10607]: How do we expect users to use this new dtype field?

File: aten/src/ATen/native/cuda/ForeachReduceOp.cu (hunk 3 above)
Comment (thomasw21) [reply aten/src/ATen/native/cuda/ForeachReduceOp.cu]: > check_fast_path_restrictions already checks that all tensors are the same dtype (which is called from can_use_fast_route, I believe), so would this check against the optional argument dtype still be necessary?

Yes, essentially you can set all tensors to fp16 and specify `dtype=torch.float`. The goal is to cast first and then compute the norm similarly to: https://pytorch.org/docs/stable/generated/torch.linalg.vector_norm.html#torch.linalg.vector_norm

> Or, if you're verifying that the tensors should match with the optional argument dtype, then checking just the first one will do right?

Actually this is true, the only issue with that is that you'd be using knowledge of what's in `can_use_fast`. I guess the same argument could be made to `has_int_or_complex` right above. You could compare just one assuming that `can_use_fast` already checks if all tensors have same dtype.

 > Maybe I'm a little confused--why are we checking that any of the tensors are the same dtype as the optional argument dtype?

That's bad implem for my end. It's supposed to be `all`. I've replace the implementation by computing the opposite (which should be any and removed the `!` in the if condition)

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 2 above)
Comment (thomasw21) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2068]: No worries
